mealie/routes/recipe/recipe_crud_routes.py

```diff
@@ -252,6 +253,11 @@ def get_all(
             if search_query.cookbook is None:
                 raise HTTPException(status_code=404, detail="cookbook not found")
 
+        if q.order_by == "random":
+            if not search_query.timestamp:
+                search_query.timestamp = datetime.now().timestamp()
```

Comment:
Instead of supplying our own here as a safeguard, we should instead use a Pydantic validation step to ensure that if the order_by is 'random' they provide a timestamp, if the client doesn't provide a timestamp, we should respond with an error code. 

https://docs.pydantic.dev/latest/usage/validators/

Reply:
I didn't know about pydantic validators, and they look very cool. But maybe overkill here? Several other methods in recipe_crud_routes just use try/except blocks to raise an HTTPException if input is invalid. Maybe that would be simpler, since it's just testing whether the timestamp exists?

Reply:
We should be doing as much validation on the edges of the application as possible. if the client provides the wrong data, we shouldn't let the request proceed and return a useful error message. That's why we're using Pydantic, for that client data validation layer.

Reply:
@hay-kot I finally had a moment to read up on validators. I saw that one use case is to set default values at validation time, with a specific example in the pydantic docs given to set a timestamp to now. So I just want to double-check your preference.

Coming in via the front end, the random_seed will get set by index.vue. But coming in from the API, random_seed is a settable parameter. You previously suggested raising a ValueError if `order_by="random"` and `random_seed = ""`. But we could instead set `random_seed = datetime.now()`, and users would be buffered from forgetting to set a seed.

Do you prefer explicit (users must always set a seed or it's an error), or implicit (failing to set a seed uses datetime.now)?

Reply:
@hay-kot I finally had a moment to read up on validators. I saw that one use case is to set default values at validation time, with a specific example in the pydantic docs given to set a timestamp to now. So I just want to double-check your preference.

Coming in via the front end, the random_seed will get set by index.vue. But coming in from the API, random_seed is a settable parameter. You previously suggested raising a ValueError if `order_by="random"` and `random_seed = ""`. But we could instead set `random_seed = datetime.now()`, and users would be buffered from forgetting to set a seed.

Do you prefer explicit (users must always set a seed or it's an error), or implicit (failing to set a seed uses datetime.now)?

Reply:
IMO, since pagination is completely broken without a seed, as an end user, if I don't supply a seed, I would want to see an error. Otherwise, if I paginate without supplying a seed, I'm getting random pages with repeats and I think the API is broken

Reply:
> Maybe be worth renaming this to randomSeed for clarity as well

I also like this, since that way it's clear that you can generate any value, not necessarily a timestamp

Reply:
Also, since this seems like it would be the only pydantic validation inside recipe_crud_routes, I'm having trouble figuring out examples to see how to get this ValidationError back to the user as an error code. Every other error code in recipe_crud_routes is a simple if/else or try/except block that raises HTTPException when there's a problem. For example, in the same recipe_crud_routes where I would be validating the random_seed:

`if search_query.cookbook is None:
          raise HTTPException(status_code=404, detail="cookbook not found")
`

So would it be fine for you to just follow the precedent in recipe_crud_routes and do the validation in a simpler way without resorting to pydantic?

Reply:
Pydantic validation errors are automatically propagated up with a clear error message.

https://fastapi.tiangolo.com/tutorial/body/#results

> if search_query.cookbook is None: raise HTTPException(status_code=404, detail="cookbook not found") 

TBH we shouldn't be throwing HTTP there anyways. Also looking at that code I'm pretty sure it's not even doing what is was intended to do. It should be checking if the cookbook_data is None and _not_ search_query.cookbook which is already guarded right above it. 

---

This or more or less what you should need

```
    @validator("random_seed", always=True, pre=True)
    def validate_random_seed(random_seed: str, values):  # type: ignore
        if values.get("order_by") == "random" and not random_seed:
            raise ValueError("Random Seed Required for Random Order")
```

Reply:
Yeah, I have basically that already and it’s just spitting the ValidationError to the command line and an Internal Server Error back to the API. Hence my confusion about propagation. But good to know this is supposed to work automagically. 

I probably have a stupid error somewhere, but my daughter just woke up so I’ll troubleshoot later.